File: 프로그래머스/2/42890. 후보키/후보키.java
import java.util.*;

class Solution {
    public int solution(String[][] relation) {
        int C = relation[0].length;
        Set<Set<Integer>> v = new HashSet<>();
        List<Set<Integer>> candidates = new ArrayList<>();
        Queue<Set<Integer>> q = new ArrayDeque<>();
        for(int i = 0; i < C; i++) {
            Set<Integer> temp = new TreeSet<>();
            temp.add(i);
            q.offer(temp);
        }
        
        while(!q.isEmpty()) {
            Set<Integer> colSet = q.poll();    
            boolean isMinColSet = true;
        
            for(Set<Integer> candidate : candidates) {
                boolean allIncluded = true;

                for(int col : candidate) {
                    if(!colSet.contains(col)) {
                        allIncluded = false;
                        break;
                    }
                }

                if(allIncluded) {
                    isMinColSet = false;
                    break;
                }
            }
            
            if(!isMinColSet) continue;
                
            
            Set<String> keySet = new HashSet<>();
            boolean isUnique = true;
            for(String[] row : relation) {
                StringBuilder sb = new StringBuilder();
                for(int c : colSet) {
                    sb.append(row[c]);
                    sb.append("/");
                }
                String key = sb.toString();
                if(keySet.contains(key)) {
                    isUnique = false;
                    break;
                }
                keySet.add(key);
            }
            
            if(isUnique) {
                candidates.add(colSet);
                continue;
            }
            
            
            for(int c = 0; c < C; c++) {
                if(colSet.contains(c)) continue;
                
                Set<Integer> nextColSet = new TreeSet<>(colSet);
                nextColSet.add(c);
                if(v.contains(nextColSet)) continue;
                
                v.add(nextColSet);
                q.offer(nextColSet);
            }
        }
        
        return candidates.size();
    }
}

File: 프로그래머스/2/42890. 후보키/후보키.cpp
#include <string>
#include <vector>
#include <queue>
#include <set>

using namespace std;

int solution(vector<vector<string>> relation) {
    int answer = 0;
    int C = relation[0].size();
    vector<set<int>> candidates;
    queue<set<int>> q;
    for(int c = 0; c < C; c++) {
        set<int> temp;
        temp.insert(c);
        q.push(temp);
    }
    
    set<set<int>> v;
    
    while(q.size() != 0) {
        set<int> cur = q.front();
        q.pop();
        
        bool isMinSet = true;
        
        for(set<int> candidate : candidates) {
            bool isAllIncluded = true;
            
            for(int c : candidate) {
                if(!cur.count(c)) {
                    isAllIncluded = false;
                    break;
                }
            }
            
            if(isAllIncluded) {
                isMinSet = false;
                break;
            }
        }
        
        if(!isMinSet) continue;
        
        bool isUnique = true;
        set<string> keySet;
        for(vector<string> row : relation) {
            string str = "";
            for(int c : cur) {
                str += row[c];
                str += "/";
            }
            
            if(keySet.count(str)) {
                isUnique = false;
                break;
            }
            keySet.insert(str);
        }
        
        if(isUnique) {
            candidates.push_back(cur);
            continue;
        }
        
        for(int c = 0; c < C; c++) {
            if(cur.count(c)) continue;
            
            set<int> nextSet;
            for(int e : cur) nextSet.insert(e);
            nextSet.insert(c);
            if(v.count(nextSet)) continue;
            
            v.insert(nextSet);
            q.push(nextSet);
        }
    }
    
    return candidates.size();
}